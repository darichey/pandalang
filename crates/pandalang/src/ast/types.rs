// TODO: syntax for polymorphic types
#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Simple(String),
    Fun(Fun),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub from: Box<Type>,
    pub to: Box<Type>,
}
