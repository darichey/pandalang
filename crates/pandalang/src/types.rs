use std::collections::HashMap;

use crate::ast::{self, Expr};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Int,
    Str,
    Fun(Box<Type>, Box<Type>),
}

impl std::fmt::Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Int => write!(f, "Int"),
            Type::Str => write!(f, "Str"),
            Type::Fun(i, o) => write!(f, "({} -> {})", i, o),
        }
    }
}

macro_rules! types {
    ($($k:expr => $v:ident),* $(,)?) => {{
        std::collections::HashMap::from([$(($k.to_string(), vec![Type::$v]),)*])
    }};
}

pub struct Context {
    types: HashMap<String, Vec<Type>>,
}

impl Context {
    pub fn new() -> Context {
        Context { types: types!() }
    }

    pub fn check(&mut self, expr: Expr) -> Result<Type, Error> {
        match expr {
            Expr::Int(_) => Ok(Type::Int),
            Expr::Str(_) => Ok(Type::Str),
            Expr::Var(ast::Var { name }) => match self.lookup(&name) {
                Some(t) => Ok(t),
                None => Err(Error::NotInScope { name }),
            },
            Expr::BinOp(ast::BinOp { left, right, kind }) => match kind {
                ast::BinOpKind::Add
                | ast::BinOpKind::Sub
                | ast::BinOpKind::Mul
                | ast::BinOpKind::Div => {
                    let lhs_t = self.check(*left)?;
                    let rhs_t = self.check(*right)?;

                    expect_t(&Type::Int, &lhs_t)?;
                    expect_t(&Type::Int, &rhs_t)?;

                    Ok(Type::Int)
                }
            },
            Expr::Let(_) => {
                panic!("Let in type checking")
            }
            Expr::Fun(ast::Fun { patt, body }) => {
                let (name, name_t) = match patt {
                    ast::Pattern::Id { name, typ } => (name, parse_type(typ)?),
                };
                self.push_type(&name, &name_t);
                let body_t = self.check(*body)?;
                self.pop_type(&name);
                Ok(Type::Fun(Box::new(name_t), Box::new(body_t)))
            }
            Expr::App(ast::App { fun, arg }) => {
                let fun_t = self.check(*fun)?;
                match fun_t {
                    Type::Fun(in_t, out_t) => {
                        let arg_t = self.check(*arg)?;
                        expect_t(&in_t, &arg_t)?;
                        Ok(*out_t)
                    }
                    _ => Err(Error::NotFunction { actual: fun_t }),
                }
            }
        }
    }

    fn lookup(&self, name: &String) -> Option<Type> {
        let ts = self.types.get(name)?;
        let t = ts.last()?;
        Some(t.clone())
    }

    fn push_type(&mut self, name: &String, t: &Type) {
        match self.types.get_mut(name) {
            Some(types) => types.push(t.clone()),
            None => {
                self.types.insert(name.clone(), vec![t.clone()]);
            }
        }
    }

    fn pop_type(&mut self, name: &String) {
        if let Some(types) = self.types.get_mut(name) {
            types.pop();
        }
    }
}

fn expect_t(expected: &Type, actual: &Type) -> Result<(), Error> {
    if expected != actual {
        Err(Error::Incompatible {
            expected: expected.clone(),
            actual: actual.clone(),
        })
    } else {
        Ok(())
    }
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Error {
    Incompatible { expected: Type, actual: Type },
    NotFunction { actual: Type },
    NotInScope { name: String },
    UnknownType { name: String },
}

fn parse_type(typ: ast::Type) -> Result<Type, Error> {
    match typ {
        ast::Type::Base { name } => match name.as_str() {
            "Int" => Ok(Type::Int),
            "Str" => Ok(Type::Str),
            _ => Err(Error::UnknownType { name }),
        },
        ast::Type::Fun(i, o) => Ok(Type::Fun(
            Box::new(parse_type(*i)?),
            Box::new(parse_type(*o)?),
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::{Context, Error, Type};
    use crate::parser;

    fn type_check(s: String) -> Result<Type, Error> {
        let mut ctx = Context {
            types: types!("x" => Int, "y" => Int, "x'" => Int, "foo" => Int, "a" => Int, "b" => Int, "c" => Int, "d" => Int, "e" => Int),
        };
        ctx.check(*parser::parse(s.as_str()).unwrap())
    }

    #[test]
    fn types() {
        insta::glob!("snapshot_inputs/**/*.panda", |path| {
            let source = type_check(std::fs::read_to_string(path).unwrap());
            insta::assert_debug_snapshot!(source);
        });
    }
}
