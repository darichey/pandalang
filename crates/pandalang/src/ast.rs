#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub patt: Pattern,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub patt: Pattern,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
        }
        .to_string()
    }
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Pattern {
    Id {
        name: String,
        // typ: Type
    },
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Base { name: String },
    Fun(Box<Type>, Box<Type>),
}
