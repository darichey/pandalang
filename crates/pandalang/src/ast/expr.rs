#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Unit,
    Bool(Bool),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
    If(If),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Bool {
    pub b: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct If {
    pub check: Box<Expr>,
    pub then: Box<Expr>,
    pub els: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eql,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
            BinOpKind::Eql => "==",
        }
        .to_string()
    }
}
