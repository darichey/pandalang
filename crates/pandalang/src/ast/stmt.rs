use super::{expr::Expr, types::Type};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Stmt {
    Let(Let),
    Declare(Declare),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub rec: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Declare {
    pub name: String,
    pub typ: Type,
}
