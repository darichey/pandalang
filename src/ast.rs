#[derive(PartialEq, Eq, Debug)]
pub enum Expr {
    Int(i64),
    Var(String),
    BinOp {
        left: Box<Expr>,
        right: Box<Expr>,
        kind: BinOpKind,
    },
    Let {
        name: String,
        value: Box<Expr>,
        body: Box<Expr>,
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}
