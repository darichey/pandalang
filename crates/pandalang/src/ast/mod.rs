use self::stmt::Stmt;

pub mod expr;
pub mod stmt;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
