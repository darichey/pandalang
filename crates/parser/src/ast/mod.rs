use self::stmt::Stmt;

pub mod expr;
pub mod stmt;
pub mod types;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
