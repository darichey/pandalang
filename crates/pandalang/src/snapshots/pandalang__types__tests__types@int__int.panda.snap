---
source: crates/pandalang/src/types.rs
expression: source
input_file: crates/pandalang/src/snapshot_inputs/int/int.panda
---
Ok(
    Int,
)
